### ISSUE TITLE
#Aave V2 Ethereum - Position Snapshots incorrect balances

### ISSUE THREAD

**Issue by Zeazy** [messari/subgraphs#1593]
In the Aave V2 Ethereum subgraph, I have noticed that the position snapshots are sometimes incorrect. Our services relies on the accuracy of this information.

**Example:**
Account `0x01d1f55d94a53a9517c07f793f35320faa0d2dcf` has many borrow events of WETH. Its first borrow event is for 300 WETH [here](https://etherscan.io/tx/0xe7f10694355de198f46faeff8b9182ff5618d6a6c6113ab2e850ea1c2dd098ac). Note this is variable debt.

The Messari position snapshots for the account record an incorrect balance of 958 variable debt WETH for the account at this TXN. What is causing the difference?

**GraphQL query used (note first borrow event is first snapshot in array):**
```
query MyQuery {
  account(id: "0x01d1f55d94a53a9517c07f793f35320faa0d2dcf") {
    positions(
      orderBy: balance
      orderDirection: asc
      where: {market: "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2", side: BORROWER}
    ) {
      market {
        id
      }
      balance
      side
      snapshots(
        orderBy: timestamp
        orderDirection: asc
      ) {
        hash
        blockNumber
        balance
      }
    }
  }
}
```

**Comment by bye43** [comment]
@dmelotik

**Comment by Zeazy** [comment]
To add further context, I've noticed that position snapshots do not update on liquidation events. These liquidations are reducing the deposit amount of the account so one would expect the position balance to update.

**Example:**
For account, `0x0825602e9d177b51af7d79acbdc68a746bc41ce4` positionSnapshots query for LINK only shows one position snapshot of 120 LINK (deposit).

However, current balance for the account is actually 0 LINK as the account has been liquidated multiple times.

**Comment by dmelotik** [comment]
Hi @Zeazy thanks for bringing this up! I am looking at your first example right now.

 A few things:
- We get the debt balance using this contract call: https://etherscan.io/address/0xF63B34710400CAd3e044cFfDcAb00a0f32E33eCf#readProxyContract#F5
- It is definitely possible that the user borrowed 300 eth and added onto an existing position which would make the balance 900.
- Here is the transaction you are referring [to](https://api.thegraph.com/subgraphs/id/QmXN7Cod92ZhstWWXxWk79GnSbkDKu4W7Bu9rSQ3cYYL9G/graphql?query=query+MyQuery+%7B%0A++borrows%28%0A++++where%3A+%7Bhash%3A+%220xe7f10694355de198f46faeff8b9182ff5618d6a6c6113ab2e850ea1c2dd098ac%22%7D%0A++%29+%7B%0A++++id%0A++++blockNumber%0A++++market+%7B%0A++++++name%0A++++%7D%0A++++amount%0A++++position+%7B%0A++++++id%0A++++++balance%0A++++++snapshots+%7B%0A++++++++blockNumber%0A++++++++balance%0A++++++%7D%0A++++%7D%0A++%7D%0A%7D). You can see this transaction opened the position, and there is a point later on where the balance is 979 after adding onto that borrow?
- It seems accurate to me

Looking at the liquidation example:
- Even though it has been liquidated it is actually not 0. The liquidations don't always perfectly capture all of the debt. Right now the position balance is like `.000006`LINK which is <$0.01 worth of LINK. The cost of gas is much more than the profit a liquidator will make so it will likely stay there unless the account adds more to their LINK borrow. See the screenshot.
- This seems to be reporting things accurately to me. Here is the [query](https://api.thegraph.com/subgraphs/id/QmXN7Cod92ZhstWWXxWk79GnSbkDKu4W7Bu9rSQ3cYYL9G/graphql?query=query+MyQuery+%7B%0A++account%28id%3A+%220x0825602e9d177b51af7d79acbdc68a746bc41ce4%22%29+%7B%0A++++liquidationCount%0A++++liquidateCount%0A++++positions+%7B%0A++++++id%0A++++++market+%7B%0A++++++++name%0A++++++%7D%0A++++++side%0A++++++balance%0A++++++snapshots%28orderBy%3A+blockNumber%2C+orderDirection%3A+asc%29+%7B%0A++++++++blockNumber%0A++++++++balance%0A++++++%7D%0A++++%7D%0A++%7D%0A%7D) I am looking at
<img width="1311" alt="Screen Shot 2023-01-12 at 11 40 35 AM" src="https://user-images.githubusercontent.com/60583747/212151739-8f66727f-0d08-4c2d-8eff-847abf852e80.png">

**Comment by Zeazy** [comment]
Hey @dmelotik thank you very much for the speedy and detailed response.

- Great thank you - that cleared it up. To confirm, [this](https://api.thegraph.com/subgraphs/id/QmXN7Cod92ZhstWWXxWk79GnSbkDKu4W7Bu9rSQ3cYYL9G/graphql?query=%0Aquery+MyQuery+%7B%0A++account%28id%3A+%220x01d1f55d94a53a9517c07f793f35320faa0d2dcf%22%29+%7B%0A++++positions%28%0A++++++orderBy%3A+balance%0A++++++orderDirection%3A+asc%0A++++++where%3A+%7BhashClosed%3A+null%7D%0A++++++%23+where%3A+%7Bmarket%3A+%220xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2%22%2C+side%3A+BORROWER%7D%0A%0A++++%29+%7B%0A++++++market+%7B%0A++++++++id%0A++++++%7D%0A++++++balance%0A++++++side%0A++++++%23+snapshots%28%0A++++++%23+++orderBy%3A+timestamp%0A++++++%23+++orderDirection%3A+asc%0A++++++%23+%29+%7B%0A++++++%23+++hash%0A++++++%23+++blockNumber%0A++++++%23+++balance%0A++++++%23+%7D%0A++++%7D%0A++%7D%0A%7D&operationName=MyQuery) query should give me all open positions (i.e. current balance) for any account.

- Oh wow - you are correct on the query you shared. What is strange is that the results in 'The GraphiQL' are completely different to the ones if you run the same query against your Aave V2 Ethereum [Playground](https://thegraph.com/explorer/subgraphs/84CvqQHYhydZzr2KSth8s1AFYpBRzUbVJXq6PWuZm9U9?view=Playground) on The Graph's website. Is this expected?

- Last question, the query you shared [here](https://api.thegraph.com/subgraphs/id/QmXN7Cod92ZhstWWXxWk79GnSbkDKu4W7Bu9rSQ3cYYL9G/graphql?query=query+MyQuery+%7B%0A++account%28id%3A+%220x0825602e9d177b51af7d79acbdc68a746bc41ce4%22%29+%7B%0A++++liquidationCount%0A++++liquidateCount%0A++++positions+%7B%0A++++++id%0A++++++market+%7B%0A++++++++name%0A++++++%7D%0A++++++side%0A++++++balance%0A++++++snapshots%28orderBy%3A+blockNumber%2C+orderDirection%3A+asc%29+%7B%0A++++++++blockNumber%0A++++++++balance%0A++++++%7D%0A++++%7D%0A++%7D%0A%7D) is very useful to get the position snapshots of an account over - time. What would make this perfect is if we could also get either the token price (usd) of the token or the account's amount (usd) of this token balance at each position snapshot. E.g. 100 balance of WETH = 200K in march, 150K in April etc. 

On other queries such as `borrows` one [here](https://api.thegraph.com/subgraphs/id/QmXN7Cod92ZhstWWXxWk79GnSbkDKu4W7Bu9rSQ3cYYL9G/graphql?query=%7B%0A++borrows%28%0A++++first%3A+1000%2C+%0A++++orderBy%3A+timestamp%2C%0A++++orderDirection%3A+asc%2C+%0A++++where%3A+%7B%0A++++++timestamp_gt%3A+1560905090%2C+%0A++++++timestamp_lte%3A+1673984742%2C+%0A++++++account%3A%220x0825602e9d177b51af7d79acbdc68a746bc41ce4%22%7D%29%0A++%7B%0A++++asset%7Bdecimals%7D+%0A++++hash+%0A++++timestamp+%0A++++blockNumber+%0A++++asset%7Bid%7D+%0A++++asset%7Bsymbol%7D+%0A++++amount+%0A++++amountUSD%0A++%7D%0A%7D%0A+) have this historical amount USD. Is there a way to get it for this PositionSnapshot query?

**Comment by dmelotik** [comment]
@Zeazy 

1. Yes, if there is a balance or `hashClosed` is not avail then the position is open.

2. Are you using the decentralized network [version](https://thegraph.com/explorer/subgraphs/84CvqQHYhydZzr2KSth8s1AFYpBRzUbVJXq6PWuZm9U9?view=Overview)? Bc we have had issues with that giving different responses from the hosted service

3. That is a longer term change we made when we upgraded the lending schema to 3.0.0 - It is available for compound V3 right now.
